### PR TITLE
Make error msg for updating resource version more clear

### PR DESCRIPTION
**Pull request by jhadvig**
The error message for conflicts in the resource versions is not clear.

### CODE REVIEW

**Comment by jhadvig** [comment]
@smarterclayton thoughts ?

**Comment by jhadvig** [comment]
I'm a RedHat employee

**Comment by smarterclayton** [comment]
LGTM

**Comment by smarterclayton** [comment]
Came up in usability from a wide range of operators - this is the most common cause of update failures on pods and nodes, so it's important to communicate clearly why the error failed.

File: pkg/registry/generic/etcd/etcd.go [hunk 1]
```diff
@@ -301,7 +301,7 @@ func (e *Etcd) Update(ctx api.Context, obj runtime.Object) (runtime.Object, bool
 		}
 		if newVersion != version {
 			// TODO: return the most recent version to a client?
-			return nil, nil, kubeerr.NewConflict(e.EndpointName, name, fmt.Errorf("the resource was updated to %d", version))
+			return nil, nil, kubeerr.NewConflict(e.EndpointName, name, fmt.Errorf("another client made an update to this object in resource version %d", version))
```
Comment (lavalamp) [inline pkg/registry/generic/etcd/etcd.go]: I don't think this text is a net win. How about something more like one of these:

* "the resource has been changed since you read it; re-read it and try your change again"
* "the resource's version has been incremented to %v; please retry"
* "changes must be made against the most recent resource version"
* "the resourceVersion you reported (%v) is out-of-date; please re-read the object and try your change again"

Comment (smarterclayton) [inline pkg/registry/generic/etcd/etcd.go]: ----- Original Message -----
> > @@ -301,7 +301,7 @@ func (e *Etcd) Update(ctx api.Context, obj
> > runtime.Object) (runtime.Object, bool
> >  		}
> >  		if newVersion != version {
> >  			// TODO: return the most recent version to a client?
> > -			return nil, nil, kubeerr.NewConflict(e.EndpointName, name,
> > fmt.Errorf("the resource was updated to %d", version))
> > +			return nil, nil, kubeerr.NewConflict(e.EndpointName, name,
> > fmt.Errorf("another client made an update to this object in resource
> > version %d", version))
> 
> I don't think this text is a net win. How about something more like one of
> these:
> 
> * "the resource has been changed since you read it; re-read it and try your
> change again"
> * "the resource's version has been incremented to %v; please retry"
> * "changes must be made against the most recent resource version"
> * "the resourceVersion you reported (%v) is out-of-date; please re-read the
> object and try your change again"

We wanted to catch the version (useful for anyone who knows what they have to do) and that the likely cause was someone else actually changed this object (to slow someone down from doing an overwrite, or to explain to a brand new user why it failed).  The first example with resource version included would be fine for me as well.

Comment (lavalamp) [inline pkg/registry/generic/etcd/etcd.go]: There's two problems with the wording:
* apiserver doesn't know if it was another client that made the conflicting change (it probably is, but we shouldn't report it as certain-- imagine a user modifying a pod and then modifying it again without re-getting it: this error will be very confusing for them.)
* "made an update to this object in resource version X" is grammatically confusing and hard to parse. And by my parsing, incorrect-- the first offending change would actually be resourceVersion X+1.

I agree we can do better than the old wording, and that making this clear is a good goal, but I don't agree that the new wording will actually help people more than it confuses them.

I actually think my last suggestion ("the resourceVersion you reported (%v) is out-of-date; please re-read the object and try your change again") is best. Maybe amended to "the resourceVersion you reported (%v) is out-of-date, the object has been modified; please re-read the object and try your change again"

Comment (smarterclayton) [inline pkg/registry/generic/etcd/etcd.go]: ----- Original Message -----
> > @@ -301,7 +301,7 @@ func (e *Etcd) Update(ctx api.Context, obj
> > runtime.Object) (runtime.Object, bool
> >  		}
> >  		if newVersion != version {
> >  			// TODO: return the most recent version to a client?
> > -			return nil, nil, kubeerr.NewConflict(e.EndpointName, name,
> > fmt.Errorf("the resource was updated to %d", version))
> > +			return nil, nil, kubeerr.NewConflict(e.EndpointName, name,
> > fmt.Errorf("another client made an update to this object in resource
> > version %d", version))
> 
> There's two problems with the wording:
> * apiserver doesn't know if it was another client that made the conflicting
> change (it probably is, but we shouldn't report it as certain-- imagine a
> user modifying a pod and then modifying it again without re-getting it: this
> error will be very confusing for them.)
> * "made an update to this object in resource version X" is grammatically
> confusing and hard to parse. And by my parsing, incorrect-- the first
> offending change would actually be resourceVersion X+1.
> 
> I agree we can do better than the old wording, and that making this clear is
> a good goal, but I don't agree that the new wording will actually help
> people more than it confuses them.
> 
> I actually think my last suggestion ("the resourceVersion you reported (%v)
> is out-of-date; please re-read the object and try your change again") is
> best. Maybe amended to "the resourceVersion you reported (%v) is
> out-of-date, the object has been modified; please re-read the object and try
> your change again"

Some of that is almost like instructions to a programmer, and also long.  Also, %v was the new resource version, not the old one, so the user could copy and paste it into whatever they're changing (in the unlikely event the conflict happens with a human or is a one off, vs the more common continual updates of the kubelets/rcs).

Comment (lavalamp) [inline pkg/registry/generic/etcd/etcd.go]: Not sure if I agree with giving the new resource version, because someone copy and pasting it is probably blowing away a change that they don't intend to.

Comment (smarterclayton) [inline pkg/registry/generic/etcd/etcd.go]: Not sure there is much value in the older value - either you set it and don't need it, or you didn't set it and so it means nothing to you.  The new value at least tells you what is on the server which doesn't require you potentially dropping context to run kubectl get foo bar.

Comment (lavalamp) [inline pkg/registry/generic/etcd/etcd.go]: Sure, I don't think that printing the old one or not is a big deal. I still think printing the new version basically enables user error; I can't think of a way to use that information safely.

**Comment by jhadvig** [comment]
@lavalamp how about:
`the resource version has been incremented to %v since you read it; re-read it and try your change again`

File: pkg/registry/generic/etcd/etcd.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/registry/generic/etcd/etcd.go]: That's fair - hadn't followed the thought through to that.

"the resourceVersion you reported (%v) is out-of-date, the object has been modified; please re-read the object and try your change again"

How about: "the object has been modified; please apply your changes to the latest version and try again"

----- Original Message -----

Comment (lavalamp) [inline pkg/registry/generic/etcd/etcd.go]: > "the object has been modified; please apply your changes to the latest version and try again"

SGTM
